core/src/main/java/io/parsingdata/metal/expression/value/Cat.java

```diff
@@ -34,12 +40,8 @@ public Cat(final ValueExpression left, final ValueExpression right) {
 
     @Override
     public Optional<Value> eval(final Value left, final Value right, final ParseState parseState, final Encoding encoding) {
-        final byte[] leftBytes = left.getValue();
-        final byte[] rightBytes = right.getValue();
-        final byte[] concatenatedBytes = new byte[leftBytes.length + rightBytes.length];
-        System.arraycopy(leftBytes, 0, concatenatedBytes, 0, leftBytes.length);
-        System.arraycopy(rightBytes, 0, concatenatedBytes, leftBytes.length, rightBytes.length);
-        return Optional.of(new Value(createFromBytes(concatenatedBytes), encoding));
+        return createFromSource(new ConcatenatedValueSource(left, right), ZERO, left.getLength().add(right.getLength()))
+            .flatMap(source -> Optional.of(new Value(source, encoding)));
```

Comment:
:tada:

Reply:
Flatmap is not necessary I think?

Reply:
Also, I do not see an Optional empty, perhaps you forgot to clean up imports? :D